### DISCUSSION TITLE
Sorting delivery methods doesn't work for Local Delivery

### DISCUSSION THREAD

**Discussion by SaeedYasin** [Shopify/function-examples#196] [Feedback: Delivery Customization API]
Here function output is defining the sorting of the shipping methods, but that is NOT what happens on the checkout. Local delivery seems to have a mind of it's own and sometimes appears randomly as the last shipping method, which is not what the sorting function specified. Is this known behaviour or issue?

**Function Stdin:**
<img width="622" alt="Screenshot 2023-03-13 at 12 46 50 AM" src="https://user-images.githubusercontent.com/26478201/224559517-d68d5899-d66c-48c0-aa74-256ab1592f27.png">

**Function Stdout:**
<img width="907" alt="Screenshot 2023-03-13 at 12 47 05 AM" src="https://user-images.githubusercontent.com/26478201/224559553-c31b1808-ce37-4769-a197-34c2c520bf08.png">

**Checkout:**
<img width="646" alt="Screenshot 2023-03-13 at 12 47 18 AM" src="https://user-images.githubusercontent.com/26478201/224559566-cd09110f-50f6-4351-b84e-f2288cb92b1d.png">

**Store name:**
https://sylab-store.myshopify.com/

**Comment by SaeedYasin** [comment]
It seems when checkout page is rendered, then function runs 2-3 times. Some of those runs only show 2 shipping methods, seems like it takes time to load Local Delivery.
<img width="913" alt="Screenshot 2023-03-13 at 1 54 40 AM" src="https://user-images.githubusercontent.com/26478201/224563136-e64e2595-d151-40cf-9f24-80f10c02199c.png">

The problem happens when it takes time to load Local Delivery option on checkout. Then even though function runs again and sorting stdout is generated correctly, Local Delivery still appears at the end of the list on checkout, which is wrong. It disregards the sorting function.

PS. There is only one function I've deployed which tries to sort shipping methods.

**Comment by RaVbaker** [comment]
Looks like the sorting is done on some shipping rates prefetch which might have only subset of delivery options.

**Comment by SaeedYasin** [comment]
Any solution for this?

**Comment by RaVbaker** [comment]
It worked on my tests. Would like to better understand the issue here but seems I miss some context about your setup @SaeedYasin. @rameshr-shopify do you have any idea for this, why could it happen this way for local delivery? 

Context for [delivery customizations API doc](https://docs.google.com/document/d/1lXkF60Yb8aql7XVziTtLpaeUm4bdL291WT301lIaVys/edit).

**Comment by rameshr-shopify** [comment]
We used to have some specific sorting logic for local delivery and instore pickup before, but it shouldn't be overriding function outputs. @SaeedYasin , are you able to share the checkout token (the url) of a checkout when you encountered this issue? We'd be able to get a better understanding of what's happening if we have that info.

**Comment by SaeedYasin** [comment]
The simplest way I am able to reproduce the issue is to just refresh the checkout shipping page a couple of times, then it will happen. Please note that this issue does happen in normal checkout flow as well, although only occasionally, so it's intermittent. When it does happen, then Local Delivery appears after a delay as compared to other shipping methods, as I described before.
 
Here is the checkout URL, please note only one function was active on this store when I did this testing,
https://sylab-store.myshopify.com/checkouts/c/8365168aa5950f292f14204273273d40/shipping

Here is a video of the checkout, I refresh the checkout page a few times and in the last refresh it shows Local Delivery at the end, although the function is set to ensure Local Delivery always appears at the top.
https://user-images.githubusercontent.com/26478201/228739792-ac807a2a-67f9-4215-afc4-66742a18495f.mov

Here is function's execution results, first I show all the function executions which happened during the capture of the video above, then I show the last function result which clearly shows that Local Delivery was set to appear at the top by the function output, but that clearly didn't happen at the checkout.
https://user-images.githubusercontent.com/26478201/228740906-2d62b097-25d1-4a87-842b-75d7a4003985.mov

**Comment by SaeedYasin** [comment]
To test it even further, I added this sorting function on a new dev store which doesn't really have anything else in it and I see different behaviour which is even stranger, it completely removed the Local Delivery option altogether from the checkout sometimes.

Here is the checkout URL, only one function (same function) was active which ensures Local Delivery is at the top,
https://sylab-dev.myshopify.com/checkouts/c/19624baf6456d9de093979fd4ee7c1ad/shipping

Here is a video of the checkout, Here you can see even the first time when I land on the checkout shipping page, it has already removed the Local Delivery option, which is wrong. I refresh the checkout page a few times and in the last one it removes the Local Delivery option again.
https://user-images.githubusercontent.com/26478201/228743197-ea4afdf9-08bc-4d8a-aebd-eee96a5af112.mov

Here is function's execution results, first I show all the function executions which happened during the capture of the video above, then I show them one by one. We can see sometimes the function input is not receiving Local Delivery as input and in the next run it does receive it. I am not sure what is causing these weird behaviours, as far as I can check I made sure settings for both stores for shipping and Local Delivery etc are identical.
https://user-images.githubusercontent.com/26478201/228744000-da660654-8506-47c0-90b1-1cb49e65d34b.mov

**Comment by SaeedYasin** [comment]
@RaVbaker @rameshr-shopify any news on this? please do let me know if you need anything else from me, thanks.

**Comment by rameshr-shopify** [comment]
I took a deeper look at this and I'm not able to reproduce this issue on my store. 

On the store https://sylab-dev.myshopify.com/, refreshing the page during the shipping step on checkout doesn't seem to result in a valid network request which seems weird to me. On a few different stores I tested on, refreshing the page on checkout always results in a network request and getting the latest shipping information (including local delivery).

@RaVbaker would you happen to know someone who'd be able to better help this issue? It doesn't seem to be anything local delivery specific which is causing this issue.

**Comment by alxwstn** [comment]
I’m running into this issue as well, and I can confirm it is unrelated to local delivery. The issue as I see it is:  While the user is on the shipping step of checkout, if any process changes the visible shipping methods, new shipping methods will often be added to the end of the list rather than placed in a proper sort order (i.e. as dictated by `move` operations in a delivery_customization function ).

My pressing use-case is this (because discount codes are likely to be added while the user is selecting shipping options): https://discord.com/channels/842813079926603828/1096090045351469156/1096098022976405534, but an easier-to-test edge case that also could affect us is customer VIP free shipping tags:

We have a delivery_customization function that hides standard shipping and shows free shipping if a customer is tagged with `enrolled_in_free_shipping`. This functions also uses move operations to ensure that visible shipping methods are sorted in ascending order of cost (free shipping at the top). If, while a customer is in the shipping stage of checkout, they are tagged with `enrolled_in_free_shipping`, and they subsequently select any shipping option (to trigger a re-render), the delivery_customization function will add the free shipping option and hide standard shipping (as expected!). However, no matter how I have fussed with move operations, the new shipping option often appears at the end of the list.

Again, this edge case (customer tagging that occurs while the customer is in checkout), is really unlikely and not a problem for our business, but it illustrates a bigger issue I’m trying to solve and appears to be a bug with checkout rendering that we can’t control. My guess is that it’s a race condition, or (if y’all are using React), it reminds me of the index-as-key list sorting issue: https://react.dev/learn/rendering-lists#keeping-list-items-in-order-with-key 

Let me know if a simplified function for issue reproduction would be useful!

**Comment by alxwstn** [comment]
@rameshr-shopify @RaVbaker any idea when y'all might have a chance to look into this? We are otherwise ready to move off of our old shipping script and onto shopify functions, but we will have to hold on rolling this out until this sorting issue is resolved

**Comment by alxwstn** [comment]
This sorting issue exists even if the store does not have a delivery customization function installed. Here are the reproduction steps using a newly-created dev store using Shopify-provided test data and no apps/customizations: https://gist.github.com/alxwstn/0ea8358375b64aaaafe73f0b42190ee9

I'll be surfacing this at office hours tomorrow for more visibility as well! Hoping to get some acknowledgement of this bug.

**Comment by TejasM** [comment]
@alxwstn I had a look your reproduce steps above and am unable to reproduce at this point. Can you check if you can still reproduce?

**Reply by alxwstn** [reply]
Yes, this is still happening:
![Screen Shot 2023-12-04 at 7 47 49 AM](https://github.com/Shopify/function-examples/assets/64493194/573e6549-6ec8-4c0f-a7cd-eca9a598255b)

**Reply by TejasM** [reply]
Can you share the development store where you are replicating this?

**Reply by alxwstn** [reply]
Absolutely! Do I need to remove password protection for you to access it? 
https://checkout-sort-issue.myshopify.com/

**Reply by linzjax** [reply]
Has there been any update on this?

**Comment by bithaolee** [comment]
I read in the dev docs that Delivery customization doesn't support local delivery and local pickup options (https://shopify.dev/docs/api/functions/reference/delivery-customization). 

It's confusing whether it's worth supporting local pickup and local delivery or not.

![image](https://github.com/Shopify/function-examples/assets/6301977/0dc56699-ba24-439b-aab2-f91dd6392b4d)
